File: p035.java
/* 
 * Solution to Project Euler problem 35
 * By Nayuki Minase
 * 
 * http://nayuki.eigenstate.org/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */


public final class p035 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p035().run());
	}
	
	
	private static final int LIMIT = Library.pow(10, 6);
	
	private boolean[] isPrime = Library.listPrimality(LIMIT - 1);
	
	
	public String run() {
		int count = 0;
		for (int i = 0; i < isPrime.length; i++) {
			if (isCircularPrime(i))
				count++;
		}
		return Integer.toString(count);
	}
	
	
	private boolean isCircularPrime(int n) {
		String s = Integer.toString(n);
		for (int i = 0; i < s.length(); i++) {
			if (!isPrime[Integer.parseInt(s.substring(i) + s.substring(0, i))])
				return false;
		}
		return true;
	}
	
}

File: p035.py
# 
# Solution to Project Euler problem 35
# by Project Nayuki
# 
# http://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 


isprime = [True] * 1000000

def compute():
	# Sieve of Eratosthenes
	isprime[0] = isprime[1] = False
	for i in range(len(isprime)):
		if isprime[i]:
			for j in range(i * i, len(isprime), i):
				isprime[j] = False
	
	ans = 0
	for i in range(len(isprime)):
		if is_circular_prime(i):
			ans += 1
	return str(ans)


def is_circular_prime(n):
	s = str(n)
	for i in range(len(s)):
		if not isprime[int(s[i : ] + s[ : i])]:
			return False
	return True


if __name__ == "__main__":
	print(compute())
